File: app/src/main/java/com/commodorethrawn/attentionapp/activity/MainActivity.java
package com.commodorethrawn.attentionapp.activity;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;

import com.commodorethrawn.attentionapp.R;
import com.commodorethrawn.attentionapp.util.ButtonUtil;
import com.google.firebase.functions.FirebaseFunctions;

import java.util.Objects;
import java.util.Timer;
import java.util.TimerTask;

public class MainActivity extends AppCompatActivity implements View.OnClickListener {

    private SharedPreferences preferences;
    private FirebaseFunctions functions;
    private Button attentionButton;
    private TextView feedbackText;
    private long lastClick;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        functions = FirebaseFunctions.getInstance();
        preferences = getSharedPreferences("attentionapp", MODE_PRIVATE);
        feedbackText = findViewById(R.id.feedbackText);
        attentionButton = findViewById(R.id.attention_button);
        attentionButton.setOnClickListener(this);
    }

    @Override
    protected void onResume() {
        super.onResume();
        boolean isFirstLaunch = !preferences.getBoolean("isSetup", false);
        if (isFirstLaunch) {
            Intent intent = new Intent(MainActivity.this, FirstLaunchActivity.class);
            startActivity(intent);
        } else if (preferences.getBoolean("isBoyfriend", false)) {
            finishAndRemoveTask();
        }
    }

    @Override
    public void onClick(View v) {
        if (System.currentTimeMillis() - lastClick > 8000) {
            ButtonUtil.doPressAnimation(this, attentionButton);
            requestAttention();
            lastClick = System.currentTimeMillis();
        } else {
            ButtonUtil.doErrorAnimation(this, attentionButton);
        }
    }

    private void requestAttention() {
        String coupleId = preferences.getString("coupleId", "");
        if (coupleId != null && !coupleId.isEmpty()) {
            functions.getHttpsCallable("requestAttention")
                    .call(coupleId)
                    .continueWith(task -> {
                        String result = (String) Objects.requireNonNull(task.getResult()).getData();
                        displayResult(result);
                        return result;
                    });
        }
    }

    private void displayResult(String result) {
        feedbackText.setText(result);
        Timer timerReset = new Timer();
        timerReset.schedule(new TimerTask() {
            @Override
            public void run() {
                feedbackText.setText("");
            }
        }, 3000);
    }

}

File: app/src/main/java/com/commodorethrawn/attentionapp/activity/MainActivity.kt
package com.commodorethrawn.attentionapp.activity;

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.TextView
import com.commodorethrawn.attentionapp.R
import com.commodorethrawn.attentionapp.util.ButtonUtil
import com.commodorethrawn.attentionapp.util.PreferenceUtil
import com.google.firebase.functions.FirebaseFunctions
import java.util.*
import kotlin.concurrent.schedule

class MainActivity : Activity(), View.OnClickListener {

    private lateinit var functions : FirebaseFunctions
    private lateinit var attentionButton : Button
    private lateinit var feedbackText : TextView
    private var lastClick : Long = -1

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        // Init vars
        PreferenceUtil.init(this)
        functions = FirebaseFunctions.getInstance()
        // Set content
        setContentView(R.layout.activity_main)
        feedbackText = findViewById(R.id.feedbackText)
        attentionButton = findViewById(R.id.attention_button)
        //  Button listener
        attentionButton.setOnClickListener(this);
    }

    override fun onResume() {
        super.onResume()
        // Start setup activity if needed
        if (!PreferenceUtil.setupComplete) {
            val intent = Intent(this, FirstLaunchActivity::class.java)
            startActivity(intent)
        } // If we're the receiver, we shouldn't be able to open the app, so we close it!
        else if (PreferenceUtil.role == PreferenceUtil.Role.RECEIVER) {
            finishAndRemoveTask()
        }
    }

    override fun onClick(button: View?) {
        // Limit how often they can do it. Firebase is a bit slow when the app has no userbase, but we still don't want spamming :)
        if (lastClick < 0 || System.currentTimeMillis() - lastClick > 8000) {
            ButtonUtil.doPressAnimation(this, attentionButton);
            requestAttention();
            lastClick = System.currentTimeMillis();
        } else {
            ButtonUtil.doErrorAnimation(this, attentionButton);
        }
    }

    private fun requestAttention() {
        val coupleId = PreferenceUtil.coupleId;
        if (coupleId.isNotEmpty()) {
            functions.getHttpsCallable("request").call(coupleId)
                    .continueWith {
                        val result = it.result?.data as String?
                        displayResult(result)
                    }
        }
    }

    private fun displayResult(result : String?) {
        feedbackText.text = result
        val timerReset = Timer()
        timerReset.schedule(3000) {
            feedbackText.text = ""
        }
    }
}